File: client-ios/Sources/CloneUI/DesignTokens.swift
import Foundation
import SwiftUI

public struct ScreenToken: Decodable {
    public let name: String?
    public let description: String?
    public let status: String?
    public let screenshot: String?
    public let hierarchy: String?
    public let directory: String?
    public let metrics: ScreenMetrics?
    public let steps: [Step]?
}

public struct ScreenMetrics: Decodable {
    public let element_count: Int?
    public let class_frequency: [String: Int]?
    public let accessibility: [AccessibilityEntry]?
}

public struct Step: Decodable {
    public let index: Int?
    public let action: String?
    public let description: String?
    public let status: String?
}

public struct AccessibilityEntry: Decodable {
    public let label: String?
    public let identifier: String?
    public let value: String?
}

public struct DesignTokens: Decodable {
    public let run_id: String?
    public let screens: [String: ScreenToken]
}

public final class TokenStore: ObservableObject {
    @Published public private(set) var tokens: DesignTokens = DesignTokens(run_id: nil, screens: [:])

    public init() {
        load()
    }

    public func load(from url: URL? = nil) {
        let fileURL = url ?? Bundle.module.url(forResource: "tokens", withExtension: "json")
        guard let fileURL else {
            print("DesignTokens: tokens.json missing")
            return
        }

        do {
            let data = try Data(contentsOf: fileURL)
            let decoder = JSONDecoder()
            tokens = try decoder.decode(DesignTokens.self, from: data)
        } catch {
            print("DesignTokens: failed to decode tokens.json -> \(error)")
        }
    }
}

File: client-android/app/src/main/java/com/example/cloneapp/DesignTokens.kt
package com.example.cloneapp

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class DesignTokens(
    @SerialName("run_id") val runId: String? = null,
    val screens: Map<String, ScreenToken> = emptyMap()
)

@Serializable
data class ScreenToken(
    val name: String? = null,
    val description: String? = null,
    val status: String? = null,
    val screenshot: String? = null,
    val hierarchy: String? = null,
    val directory: String? = null,
    val metrics: ScreenMetrics? = null,
    val steps: List<Step>? = null
)

@Serializable
data class ScreenMetrics(
    @SerialName("element_count") val elementCount: Int? = null,
    @SerialName("class_frequency") val classFrequency: Map<String, Int>? = null,
    val accessibility: List<AccessibilityEntry>? = null
)

@Serializable
data class AccessibilityEntry(
    val label: String? = null,
    val identifier: String? = null,
    val value: String? = null
)

@Serializable
data class Step(
    val index: Int? = null,
    val action: String? = null,
    val description: String? = null,
    val status: String? = null
)

object TokenStore {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun load(context: Context): DesignTokens = withContext(Dispatchers.IO) {
        val input = context.assets.open("tokens.json")
        val bytes = input.use { it.readBytes() }
        json.decodeFromString(DesignTokens.serializer(), bytes.decodeToString())
    }
}
